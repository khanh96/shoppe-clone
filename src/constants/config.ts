const config = {
  baseURL: 'https://api-ecom.duthanhduoc.com/',
  maxSizeUploadAvatar: 1048576 // 1MB (upload ảnh tính theo bytes)
}
export default config
