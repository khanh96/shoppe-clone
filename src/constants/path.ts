export const path = {
  home: '/',
  login: '/login',
  register: '/register',
  logout: '/logout',
  user: '/user',
  userAccount: '/user/account',
  profile: '/user/account/profile',
  changePassword: '/user/account/password',
  historyPurchase: '/user/purchase',
  productDetail: ':nameId',
  cart: '/cart'
} as const
