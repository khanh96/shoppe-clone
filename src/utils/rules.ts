import type { RegisterOptions, UseFormGetValues } from 'react-hook-form'
import { text } from 'stream/consumers'
import * as yup from 'yup'

type Rules = {
  [key in 'email' | 'password' | 'confirm_password']?: RegisterOptions
}

export const getRules = (getValues?: UseFormGetValues<any>): Rules => ({
  email: {
    required: {
      value: true,
      message: 'Trường Email phải bắt buộc'
    },
    maxLength: {
      value: 160,
      message: 'Độ dài email từ 5 - 160 ký tự'
    },
    minLength: {
      value: 5,
      message: 'Độ dài email từ 5 - 160 ký tự'
    },
    pattern: {
      value: /^\S+@\S+\.\S+$/,
      message: 'Email không đúng định dạng'
    }
  },
  password: {
    required: {
      value: true,
      message: 'Trường Password phải bắt buộc'
    },
    maxLength: {
      value: 160,
      message: 'Độ dài Password từ 6 - 160 ký tự'
    },
    minLength: {
      value: 6,
      message: 'Độ dài Password từ 6 - 160 ký tự'
    }
  },
  confirm_password: {
    required: {
      value: true,
      message: 'Trường Confirm Password phải bắt buộc'
    },
    maxLength: {
      value: 160,
      message: 'Độ dài Confirm Password từ 6 - 160 ký tự'
    },
    minLength: {
      value: 6,
      message: 'Độ dài Confirm Password từ 6 - 160 ký tự'
    },
    validate:
      typeof getValues === 'function'
        ? (value) => {
            if (getValues('password') === value) {
              return true
            }
            return 'Nhập lại Confirm Password không khớp'
          }
        : undefined
  }
})

function testPriceMinMax(this: yup.TestContext<yup.AnyObject>) {
  const { price_max, price_min } = this.parent as { price_min: string; price_max: string }
  if (price_min !== '' && price_max !== '') {
    return Number(price_max) >= Number(price_min)
  }
  return price_max !== '' || price_min !== ''
}

// Sử dụng schema để validate

export const schema = yup.object({
  email: yup
    .string()
    .required('Trường Email phải bắt buộc')
    .email('Email không đúng định dạng')
    .min(5, 'Độ dài email từ 5 - 160 ký tự')
    .max(160, 'Độ dài email từ 5 - 160 ký tự'),
  password: yup
    .string()
    .required('Trường Password phải bắt buộc')
    .min(6, 'Độ dài password từ 6 - 160 ký tự')
    .max(160, 'Độ dài password từ 6 - 160 ký tự'),
  confirm_password: yup
    .string()
    .required('Trường Confirm Password phải bắt buộc')
    .min(6, 'Độ dài confirm_password từ 6 - 160 ký tự')
    .max(160, 'Độ dài confirm_password từ 6 - 160 ký tự')
    .oneOf([yup.ref('password')], 'Nhập lại Confirm Password không khớp'),
  price_min: yup.string().test({
    name: 'price-not-allowed',
    message: 'Vui lòng điền khoảng giá phù hợp',
    test: testPriceMinMax
  }),
  price_max: yup.string().test({
    name: 'price-not-allowed',
    message: 'Vui lòng điền khoảng giá phù hợp',
    test: testPriceMinMax
  }),
  name: yup.string().required()
})

export const userSchema = yup.object({
  name: yup.string().max(160, 'Độ dài tối đa name là 160 ký tự'),
  phone: yup.string().max(20, 'Độ dài tối đa phone là 20 ký tự'),
  address: yup.string().max(160, 'Độ dài tối đa address là 160 ký tự'),
  avatar: yup.string().max(1000, 'Độ dài tối đã avatar là 1000 ký tự'),
  date_of_birth: yup.date().max(new Date(), 'Hãy chọn một ngày trong quá khứ'),
  new_password: yup.string().min(6, 'Độ dài password từ 6 - 160 ký tự').max(160, 'Độ dài password từ 6 - 160 ký tự'),
  password: schema.fields['password'], // kế thừa rules trong yup
  confirm_password: schema.fields['confirm_password']
})

// export type dùng schema
export const registerSchema = schema.pick(['email', 'password', 'confirm_password'])
export type RegisterSchemaType = yup.InferType<typeof registerSchema>

// kế thừa type của registerSchema dùng omit
export const loginSchema = schema.pick(['email', 'password'])
export type LoginSchemaType = yup.InferType<typeof loginSchema>

export const priceRangeSchema = schema.pick(['price_min', 'price_max'])
export type PriceRangeSchemaType = yup.InferType<typeof priceRangeSchema>

export const nameSchema = schema.pick(['name'])
export type NameSchemaType = yup.InferType<typeof nameSchema>

export const profileSchema = userSchema.pick(['name', 'phone', 'address', 'date_of_birth'])
export type UserSchemaType = yup.InferType<typeof userSchema>

// kế thừa type của registerSchema dùng pick
//export const emailASchema = registerSchema.pick(['email','password'])
// export type EmailAType = yup.InferType<typeof emailASchema>
