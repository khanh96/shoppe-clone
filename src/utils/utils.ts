import axios, { AxiosError } from 'axios'
import config from 'src/constants/config'
import HttpStatusCode from 'src/constants/httpStatusCode.enum'
import userImage from 'src/assets/images/no-avatar.png'
/**
 * Phương pháp "type predicate" dùng để thu hẹp kiểu của một biến
 * ✅ Đầu tiên chúng ta sẽ khai báo một function check kiểm tra cấu trúc về mặc logic javascript
 * ✅ Tiếp theo chúng ta thêm `parameterName is Type` làm kiểu return của function thay vì boolean
 * ✅ Khi dùng function kiểu tra kiểu này, ngoài việc kiểm tra về mặc logic cấu trúc, nó còn chuyển kiểu
 *
 * So sánh với phương pháp ép kiểu "Type Assertions" thì ép kiểu chúng giúp chúng ta đúng về mặc Type, chưa chắc về logic
 *
 */

export function isAxiosError<T>(error: unknown): error is AxiosError<T> {
  // eslint-disable-next-line import/no-named-as-default-member
  return axios.isAxiosError(error)
}

export function isAxiosUnprocessableEntity<FormError>(error: unknown): error is AxiosError<FormError> {
  return isAxiosError(error) && error.response?.status === HttpStatusCode.UnprocessableEntity
}

export function formatCurrency(currency: number) {
  return new Intl.NumberFormat('en-DE').format(currency)
}

export function formatNumberToSocialStyle(value: number) {
  return new Intl.NumberFormat('en', { notation: 'compact', minimumFractionDigits: 1 }).format(value)
}

export function rateSale(original: number, sale: number) {
  return Math.round(((original - sale) / original) * 100) + '%'
}

export const removeSpecialCharacter = (str: string) =>
  // eslint-disable-next-line no-useless-escape
  str.replace(/!|@|%|\^|\*|\(|\)|\+|\=|\<|\>|\?|\/|,|\.|\:|\;|\'|\"|\&|\#|\[|\]|~|\$|_|`|-|{|}|\||\\/g, '')

export const generateNameId = ({ name, id }: { name: string; id: string }) => {
  //: /\s/g => ký tự dấu spaces
  return removeSpecialCharacter(name).replace(/\s/g, '-') + `-id-${id}`
}

export const getIdFromNameId = (nameId: string) => {
  const arr = nameId.split('-id-')
  return arr[arr.length - 1]
}

export const getAvatarUrl = (avatarName?: string) => (avatarName ? `${config.baseURL}images/${avatarName}` : userImage)
