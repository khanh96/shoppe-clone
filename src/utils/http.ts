import axios, { AxiosError, AxiosInstance, InternalAxiosRequestConfig } from 'axios'
import { toast } from 'react-toastify'
import HttpStatusCode from 'src/constants/httpStatusCode.enum'
import {
  clearLS,
  getAccessTokenFromLS,
  getRefreshTokenFromLS,
  setAccessTokenToLS,
  setProfileToLS,
  setRefreshTokenToLS
} from './auth'
import { AuthResponse, ErrorExpireTokenResponse, RefreshTokenResponse } from 'src/types/auth.type'
import { path } from 'src/constants/path'
import config from 'src/constants/config'
import { isAxiosExpiredTokenError, isAxiosUnauthorizedError, redirectToPageNotFound } from './utils'
import { URL_LOGIN, URL_LOGOUT, URL_REFRESH_ACCESS_TOKEN, URL_REGISTER } from 'src/apis/auth.api'
import { ErrorResponseApi } from 'src/types/utils.type'

export class Http {
  instance: AxiosInstance
  private accessToken: string
  private refreshToken: string
  private refreshTokenRequest: Promise<string> | null
  constructor() {
    // Tạo biến và lưu accessToken từ localStorage để lúc lấy accessToken sẽ lấy từ RAM nhanh hơn là lấy từ bộ nhớ (localStorage)
    this.accessToken = getAccessTokenFromLS()
    this.refreshToken = getRefreshTokenFromLS()
    this.refreshTokenRequest = null
    this.instance = axios.create({
      baseURL: config.baseURL,
      timeout: 10000,
      headers: {
        'Content-Type': 'application/json',
        'expire-access-token': 60 * 60, // 1h
        'expire-refresh-token': 60 * 60 * 24 * 1000 // 1 ngay
      }
    })
    this.instance.interceptors.request.use(
      (config) => {
        if (this.accessToken) {
          config.headers.authorization = this.accessToken
          return config
        }
        return config
      },
      (error) => {
        return Promise.reject(error)
      }
    )
    this.instance.interceptors.response.use(
      (response) => {
        const { url } = response.config
        if (url === URL_LOGIN || url === URL_REGISTER) {
          const data = response.data as AuthResponse
          this.accessToken = data.data.access_token
          this.refreshToken = data.data.refresh_token
          setAccessTokenToLS(this.accessToken)
          setRefreshTokenToLS(this.refreshToken)
          setProfileToLS(data.data.user)
        } else if (url === URL_LOGOUT) {
          this.accessToken = ''
          this.refreshToken = ''
          clearLS()
        }
        return response
      },
      (error: AxiosError) => {
        // Các lỗi không phải là 422 vaf 401 thì mới nhảy vào đấy
        if (
          ![HttpStatusCode.UnprocessableEntity, HttpStatusCode.Unauthorized].includes(error.response?.status as number)
        ) {
          const data: any | undefined = error.response?.data
          const message = data?.message || error.message
          toast.error(message)
        }
        // Lỗi Unauthorized (401) có rất nhiều trường hợp
        // - Token không đúng
        // - Không truyền token
        // - Token hết hạn

        // Nếu là lỗi 401
        if (isAxiosUnauthorizedError<ErrorResponseApi<{ message: string; name: string }>>(error)) {
          const config = error.response?.config as InternalAxiosRequestConfig
          const { url } = config
          // Trường hợp token hết hạn và request đó không phải là của request refresh token
          // thì chúng ta mới tiến hành gọi refresh token
          if (isAxiosExpiredTokenError<ErrorExpireTokenResponse>(error) && url !== URL_REFRESH_ACCESS_TOKEN) {
            // Hạn chế gọi 2 lần handleRefreshToken
            this.refreshTokenRequest = this.refreshTokenRequest
              ? this.refreshTokenRequest
              : this.handleRefreshToken().finally(() => {
                  // giữ refreshTokenRequest trong 10s cho những request tiếp theo nếu có 401 thì dùng.
                  setTimeout(() => {
                    this.refreshTokenRequest = null
                  }, 10000)
                })
            return this.refreshTokenRequest.then((access_token) => {
              config.headers.authorization = access_token
              return this.instance(config)
            })
          }
          // Còn những trường hợp như token không đúng
          // không tuyển token,
          // token hết hạn nhưng gọi refresh token bị fail thì tiến hành xóa local storage và toast message
          clearLS()
          this.accessToken = ''
          this.refreshToken = ''
          toast.error(error.response?.data.data?.message || error.response?.data.message)
        }
        if (
          error.response?.status === HttpStatusCode.NotFound ||
          error.response?.status === HttpStatusCode.BadRequest
        ) {
          redirectToPageNotFound()
        }
        return Promise.reject(error)
      }
    )
  }
  private handleRefreshToken() {
    return this.instance
      .post<RefreshTokenResponse>(URL_REFRESH_ACCESS_TOKEN, { refresh_token: this.refreshToken })
      .then((res) => {
        const { access_token } = res.data.data
        setAccessTokenToLS(access_token)
        this.accessToken = access_token
        return access_token
      })
      .catch((error) => {
        clearLS()
        this.accessToken = ''
        this.refreshToken = ''
        throw error
      })
  }
}

const http = new Http().instance

export default http
