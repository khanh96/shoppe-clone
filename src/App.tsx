import { useContext, useEffect } from 'react'
import { HelmetProvider } from 'react-helmet-async'
import { ToastContainer } from 'react-toastify'
import { useNavigate } from 'react-router-dom'
import './App.css'
import useRouterElement from './routers/useRouterElement'
import 'react-toastify/dist/ReactToastify.css'
import { LocalStorageEventTarget } from './utils/auth'
import { AppContext } from './contexts/app.context'
import { PageNotFoundEventTarget } from './utils/utils'
import { path } from './constants/path'

import ErrorBoundary from './components/ErrorBoundary/ErrorBoundary'

const helmetContext = {}

/**
 * Khi url thay đổi thì các component nào dùng các hook như
 * useRoutes, useParmas, useSearchParams,...
 * sẽ bị re-render.
 * Ví dụ component `App` dưới đây bị re-render khi mà url thay đổi
 * vì dùng `useRouteElements` (đây là customhook của `useRoutes`)
 */

function App() {
  console.log('APP')
  const routeElements = useRouterElement()
  // const navigate = useNavigate()
  const { resetAuth } = useContext(AppContext)

  useEffect(() => {
    LocalStorageEventTarget.addEventListener('clearLS', () => {
      resetAuth()
    })

    return () => {
      LocalStorageEventTarget.removeEventListener('clearLS', () => {
        resetAuth()
      })
    }
  }, [resetAuth])

  // useEffect(() => {
  //   PageNotFoundEventTarget.addEventListener('pageNotFound', () => {
  //     navigate({ pathname: path.notFound })
  //   })
  //   return () => {
  //     LocalStorageEventTarget.removeEventListener('pageNotFound', () => {
  //       navigate({ pathname: path.notFound })
  //     })
  //   }
  // }, [navigate])ß

  return (
    <>
      <ErrorBoundary>
        <HelmetProvider context={helmetContext}>
          {routeElements}
          <ToastContainer />
        </HelmetProvider>
      </ErrorBoundary>
    </>
  )
}

export default App
