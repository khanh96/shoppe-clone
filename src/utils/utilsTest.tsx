import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { screen, waitFor, type waitForOptions } from '@testing-library/dom'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { BrowserRouter } from 'react-router-dom'
import App from 'src/App'
import { AppContextProvider, getInitialAppContext } from 'src/contexts/app.context'
import { expect } from 'vitest'

export const delay = (time: number) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(true)
    }, time)
  })
}

export const logScreen = async (
  body: HTMLElement = document.body.parentElement as HTMLElement,
  options?: waitForOptions
) => {
  const { timeout = 1000 } = options || {}
  await waitFor(
    async () => {
      expect(await delay(timeout - 100)).toBe(true)
    },
    {
      ...options,
      timeout: timeout
    }
  )

  screen.debug(body, 999999)
}

// Custom config queryClient
const createWrapperReactQuery = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false // Khi call api bị lỗi thì sẽ không gọi lại
      },
      mutations: {
        retry: false
      }
    },
    // Tắt log error khi test.
    logger: {
      log: console.log,
      warn: console.warn,
      error: () => null
    }
  })
  const Provider = ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  )
  return Provider
}

const ProviderReactQuery = createWrapperReactQuery()

export const renderWithRouter = ({ router = '/' } = {}) => {
  const user = userEvent.setup()
  window.history.pushState({}, 'Test page', router)
  console.log('=======>', router)
  const defaultAppContext = getInitialAppContext()
  return {
    ...render(
      <ProviderReactQuery>
        <AppContextProvider defaultValue={defaultAppContext}>
          <App />
        </AppContextProvider>
      </ProviderReactQuery>,
      {
        wrapper: BrowserRouter
      }
    ),
    user: user
  }
}
