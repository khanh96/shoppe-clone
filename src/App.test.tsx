import { describe, expect, test } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react' // sử dụng thư viện này để render trên nodejs. Vì reactjs render trên client
import userEvent from '@testing-library/user-event'
import matchers from '@testing-library/jest-dom/matchers'
import { BrowserRouter, MemoryRouter } from 'react-router-dom'
import App from './App'
import { logScreen, renderWithRouter } from './utils/utilsTest'
import { path } from './constants/path'

expect.extend(matchers)

describe('APP', () => {
  test('Test render app and navigate page', async () => {
    render(<App />, { wrapper: BrowserRouter })
    const { user } = renderWithRouter()

    /**
     * waitFor sẽ run callback 1 vài lần
     * cho đến khi hết timeout hoặc expect pass
     * số lần run phụ thuộc vào timeout và interval
     * mặc định: timeout = 1000ms và interval = 50ms
     */

    // TEST: Verify vào đúng trang chủ
    await waitFor(() => {
      expect(document.querySelector('title')?.textContent).toBe('Homepage | Shoppe Clone')
    })
    // xác minh nội dung trang cho tuyến đường dự kiến sau khi điều hướng
    await user.click(screen.getByText(/login/i))
    // Verify chuyển sang trang login
    await waitFor(() => {
      // Tìm text có trong trang web k?
      expect(screen.queryByText(/Bạn mới biết đến Shopee?/i)).toBeInTheDocument()
      expect(document.querySelector('title')?.textContent).toBe('Login | Shoppe Clone')
    })

    // Cú pháp debug unit test
    // screen.debug(document.body.parentElement as HTMLElement, 99999)
  })
  // TEST: người dùng có về được trang 404 not found khi vào 1 url không tồn tại
  test('Test về trang not found', async () => {
    const badRoute = '/abc/xyz'
    renderWithRouter({ router: badRoute })
    // render(
    //   <MemoryRouter initialEntries={[badRoute]}>
    //     <App />
    //   </MemoryRouter>
    // )
    await waitFor(() => {
      expect(screen.getByText(/Page Not Found/i)).toBeInTheDocument()
    })

    // await logScreen()
  })
  // TEST: test người dùng nhập url vào 1 trang bất kỳ vd: nhập vào url register
  test('Test render register', async () => {
    renderWithRouter({ router: path.register })
    // window.history.pushState({}, 'Test page register', path.register)
    // render(<App />, {
    //   wrapper: BrowserRouter
    // })
    // render(
    //   <MemoryRouter initialEntries={[path.register]}>
    //     <App />
    //   </MemoryRouter>
    // )
    await waitFor(() => {
      expect(screen.getByText(/Bạn đã có tài khoản Shopee?/i)).toBeInTheDocument()
    })

    // await logScreen()
  })
})
