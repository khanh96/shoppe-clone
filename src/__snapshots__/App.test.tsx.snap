// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`APP > Test render app and navigate page 1`] = `
<body>
  <div>
    <div
      class="Toastify"
    />
  </div>
</body>
`;
